### PR TITLE
Fix issues loading nested associations

### LINKED ISSUE #8485: Translate association when it is a belongsToMany (Plugin)
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.2.4

### Actual Behavior
I can not translate the contents of association when it is a belongsToMany

Posts Table
```php
$this->belongsToMany('Administrator.Fields', [ 'through' => 'Administrator.FieldsPosts', ]);
```

Fields Table
```php
$this->belongsToMany('Administrator.Posts', [ 'through' => 'Administrator.FieldsPosts']);
```

FieldsPosts Table

```php
$this->addBehavior('Translate', ['fields' => ['value']]);
$this->belongsTo('Administrator.Posts');
$this->belongsTo('Administrator.Fields');
```

BD FieldsPosts
```
ID - field_id - post_id - value - created - modified
```

PostsController
```ph
$post = $this->Posts->get($id,['contain'=>['Fields']]);
```

Error:
```
Fields is not associated with .FieldsPosts_value_translation
```

```
Notice (8): Undefined index: aliasPath [CORE\src\ORM\Association.php, line 811]
```

### PR DESCRIPTION
**Pull request by lorenzo**
This builds on top of #9191 to fix #8485. Had to target 3.3 as changes in the API were required.

I like how this ended up at the end, as it allowed me to remove a few dirty hacks that were still around.

### CODE REVIEW

File: src/ORM/Association/BelongsToMany.php [hunk 1]
```diff
@@ -1278,12 +1256,26 @@ protected function _buildQuery($options)
 
         // Ensure that association conditions are applied
         // and that the required keys are in the selected columns.
+
+        $tempName = $this->_name . '_CJoin';
+        $schema = $assoc->schema();
+        $joinFields = $types = [];
+
+        foreach ($schema->typeMap() as $f => $type ) {
```
Comment (lorenzo) [inline src/ORM/Association/BelongsToMany.php]: huh?

Comment (lorenzo) [inline src/ORM/Association/BelongsToMany.php]: Huh?

Comment (markstory) [inline src/ORM/Association/BelongsToMany.php]: It doesn't like `$type ) {` it wants `$type) {`

File: src/ORM/Association/BelongsToMany.php [hunk 2]
```diff
@@ -411,20 +411,6 @@ protected function _appendNotMatching($query, $options)
     }
 
     /**
-     * {@inheritDoc}
-     */
-    public function transformRow($row, $nestKey, $joined)
```
Comment (markstory) [inline src/ORM/Association/BelongsToMany.php]: Nice :+1:

File: src/ORM/Association/BelongsToMany.php (hunk 1 above)
Comment (lorenzo) [inline src/ORM/Association/BelongsToMany.php]: ah!

File: src/ORM/Association/BelongsToMany.php [hunk 3]
```diff
@@ -1278,12 +1256,26 @@ protected function _buildQuery($options)
 
         // Ensure that association conditions are applied
         // and that the required keys are in the selected columns.
+
+        $tempName = $this->_name . '_CJoin';
+        $schema = $assoc->schema();
```
Comment (markstory) [inline src/ORM/Association/BelongsToMany.php]: Which table does this end up proxying onto?

Comment (markstory) [inline src/ORM/Association/BelongsToMany.php]: Should we make a local reference to the `target`? There are a few uses of it now.

File: src/ORM/EagerLoadable.php [hunk 4]
```diff
@@ -86,6 +86,14 @@ class EagerLoadable
     protected $_forMatching;
 
     /**
+     * The property name where the association result should be nested
+     * in the result.
+     *
+     * @var string
+     */
+    protected $_targetProperty;
```
Comment (markstory) [inline src/ORM/EagerLoadable.php]: How does this interact with `propertyPath`?

File: src/ORM/ResultSet.php [hunk 5]
```diff
@@ -380,12 +380,14 @@ protected function _calculateColumnMap($query)
         $map = [];
         foreach ($query->clause('select') as $key => $field) {
             $key = trim($key, '"`[]');
-            if (strpos($key, '__') > 0) {
-                $parts = explode('__', $key, 2);
-                $map[$parts[0]][$key] = $parts[1];
-            } else {
+
+            if (strpos($key, '__') <= 0) {
```
Comment (markstory) [inline src/ORM/ResultSet.php]: How does the alias get leading `__` on it?

File: src/ORM/EagerLoadable.php (hunk 4 above)
Comment (lorenzo) [inline src/ORM/EagerLoadable.php]: `propertyPath` is the way up to the container of `targetProperty`. For example

```php
$article->author->profile->company;
```

Would have a `propertyPath` of `author.profile` and a `targetProperty` of `company`

Comment (markstory) [inline src/ORM/EagerLoadable.php]: We should put that in the API docs :smile:

File: src/ORM/ResultSet.php (hunk 5 above)
Comment (lorenzo) [inline src/ORM/ResultSet.php]: I just refactored the `if..else` to have no `else`, no changes in the logic where made. To answer your question, the `__` comes from the method `aliasField()` which produces `TableName__fieldName`

File: src/ORM/EagerLoadable.php (hunk 4 above)
Comment (lorenzo) [inline src/ORM/EagerLoadable.php]: I can do that

File: src/ORM/ResultSet.php (hunk 5 above)
Comment (markstory) [inline src/ORM/ResultSet.php]: Ok.

**Comment by markstory** [comment]
Thanks for updating the docs :+1:

File: src/ORM/Association/BelongsToMany.php (hunk 3 above)
Comment (lorenzo) [inline src/ORM/Association/BelongsToMany.php]: What do you mean?

Comment (markstory) [inline src/ORM/Association/BelongsToMany.php]: We directly/indirectly call `$assoc->target()` ~3 times in this method. Should we save some function calls by doing `$target = $assoc->target()`?

Comment (lorenzo) [inline src/ORM/Association/BelongsToMany.php]: `target()` is called twice, but they are actually different target tables

Comment (markstory) [inline src/ORM/Association/BelongsToMany.php]: My mistake. I just saw multiple calls and thought we could make fewer.

File: src/ORM/Association/BelongsToMany.php [hunk 6]
```diff
@@ -1278,11 +1256,26 @@ protected function _buildQuery($options)
 
         // Ensure that association conditions are applied
         // and that the required keys are in the selected columns.
+
+        $tempName = $this->_name . '_CJoin';
```
Comment (markstory) [inline src/ORM/Association/BelongsToMany.php]: Is there anything special about this name or are you just trying to ensure uniqueness?

Comment (lorenzo) [inline src/ORM/Association/BelongsToMany.php]: Trying to make it unique.

File: src/ORM/EagerLoader.php [hunk 7]
```diff
@@ -631,7 +632,8 @@ public function associationsMap($table)
                     'canBeJoined' => $canBeJoined,
                     'entityClass' => $instance->target()->entityClass(),
                     'nestKey' => $canBeJoined ? $assoc : $meta->aliasPath(),
-                    'matching' => $forMatching !== null ? $forMatching : $matching
+                    'matching' => $forMatching !== null ? $forMatching : $matching,
+                    'targetProperty' => $meta->targetProperty()
```
Comment (markstory) [inline src/ORM/EagerLoader.php]: Not entirely related to this change, but why do we map the `EagerLoadable` instances into arrays? Could we just pass the objects around? Is that possibly an artifact of `EagerLoadable` being added late-ish in the development of the ORM?

**Comment by markstory** [comment]
Looks good. I understand the reasons for the `CJoin` aliasing better now.

**Pull request merged**